File: Data Structures & Algorithms/interleaving-string/submission-1.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        memo = {}

        def dfs(i: int, j: int, k: int):
            if memo.get((i,j)) is not None:
                return memo[(i, j)]

            if k == len(s3):
                return (i == len(s1) and (j == len(s2)))
            
            if i < len(s1) and s1[i] == s3[k]:
                if dfs(i + 1, j, k + 1): return True

            if j < len(s2) and s2[j] == s3[k]:
                if dfs(i, j + 1, k + 1): return True

            memo[(i, j)] = False
            return False

        return dfs(0, 0, 0)

File: Data Structures & Algorithms/coin-change/submission-1.cpp
class Solution {
int output;
vector<int> memo; // memo[sum] = min coins used to reach sum

void backtrack(int &coin_sum, int &num_coins, const vector<int> &coins, int target){
    if (coin_sum == target){
        output = min(num_coins, output);
        return;
    }
    if (coin_sum > target) return;
    if (num_coins >= output) return;
    if (memo[coin_sum] != -1 && memo[coin_sum] <= num_coins)
            return;
    memo[coin_sum] = num_coins;
    for (auto coin: coins){
        coin_sum += coin;
        num_coins++;
        backtrack(coin_sum, num_coins, coins, target);
        coin_sum -= coin;
        num_coins--;
    }
}


public:
    int coinChange(vector<int>& coins, int amount) {
        output = INT_MAX;
        memo.assign(amount + 1, -1);
        int coin_sum = 0, num_coins = 0;
        backtrack(coin_sum, num_coins, coins, amount);

        return (output == INT_MAX) ? -1 : output;
    }
};
